Storage abstraction enhancement

lukasaric opened the pull request:
This PR improves storage abstraction.

- [x] Add unified `IStorage` interface
- [x] Defined return/response types
- [x] Remove passing magic `options` param 🤡 
- [x] Use built-in promise based `fs` in filesystem storage implementation

lukasaric committed:
Unified storage methods within the IStorage

lukasaric committed:
Changed write stream creation interface

lukasaric committed:
Cleanup

lukasaric committed:
Exposed createStorage fn

lukasaric committed:
Added missing semicolons

lukasaric committed:
Init storage provider in main

lukasaric committed:
Fixed exists method :wrench:

lukasaric committed:
Filesystem methods cleanup :nail_care:

lukasaric committed:
Removed mode param

lukasaric committed:
Amazon methods cleanup :art:

lukasaric committed:
Used FileListResponse type for list files

lukasaric committed:
Removed unnecessary validation

- Validation is covered when creating storage config

lukasaric committed:
Add extra provider config extraction type check

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/storage-abstraction-enhancement

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/storage-abstraction-enhancement

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/storage-abstraction-enhancement

MiroDojkic commented on server/shared/storage/amazon.ts:
```diff
@@ -0,0 +1,131 @@
+  #client: S3;
+
+  constructor(config: Config) {
+    const amazonConfig = config.storage.amazon as S3Config;
```
Why do we need to cast the type here?

MiroDojkic commented on server/shared/storage/amazon.ts:
```diff
@@ -0,0 +1,131 @@
+      signatureVersion: 'v4',
+      apiVersion: '2006-03-01',
+      maxRetries: 3
+    } as ClientConfiguration;
```
Why do we need to cast here?

MiroDojkic commented on server/shared/storage/amazon.ts:
```diff
@@ -0,0 +1,131 @@
+    this.#client = new S3(s3Config);
+  }
+
+  private path(...segments): string {
```
Can we stick to the following order of the class members:
1. public fields
2. private fields
3. constructor
4. public methods
5. private methods

MiroDojkic commented on server/shared/storage/amazon.ts:
```diff
@@ -0,0 +1,131 @@
+
+  private path(...segments): string {
+    segments = [this.#bucket, ...segments];
+    return path.join(...segments);
```
Does this work (seems simpler)?
```suggestion
    return path.join(this.#bucket, ...segments);
```

MiroDojkic commented on server/shared/storage/index.ts:
```diff
@@ -1,58 +1,2 @@
-}
-
-export default Storage;
+export { default as amazon } from './amazon';
+export { default as filesystem } from './filesystem';
```
It seems that we could nicely encapsulate instantiation logic with a factory function that depends on config here?

lukasaric replied:
👍

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/storage-abstraction-enhancement

lukasaric committed:
Removed unnecessary promise resolve

lukasaric committed:
Use fs exists method

lukasaric committed:
Removed unnecessary castin & path resolve fix

lukasaric committed:
Class members order :eyes:

lukasaric committed:
Init storage within the createStorage fn

lukasaric committed:
Set proper return type on write stream method

lukasaric committed:
Set proper return type for file exists method

lukasaric committed:
Delete response type & methods enhancement

lukasaric committed:
Remove unused not found msg

lukasaric committed:
Changed file list response type to have path only

lukasaric committed:
Remove Response interface

lukasaric committed:
Set raw prop to be Buffer :wrench:

lukasaric committed:
Added AWSError type check

lukasaric committed:
Add system err interface :eyes:

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/storage-abstraction-enhancement

lukasaric committed:
Defined responses as types

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/storage-abstraction-enhancement

lukasaric committed:
Encoded copy source file name

ikovac commented on server/shared/storage/amazon.ts:
```diff
@@ -0,0 +1,123 @@
+    this.#client = new S3(s3Config);
+  }
+
+  // API docs: http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#getObject-property
```
Do we need API docs comment for every storage method? If docs are useful can we just link the main docs page instead of linking docs for every method we use (some of them even twice)?

lukasaric replied:
I definitely agree with the statement that some of them that are repeating, but still it can be handy when somebody wants to change the implementation of the method. @MiroDojkic thoughts?

MiroDojkic replied:
@malizeus I think these are handy, but I also understand Ivo's point. I'm ok with both a single link or link for each method.

MiroDojkic replied:
@malizeus link seems to point to the latest API docs though, and we're locking the version that we use. Should we change link(s) to point to a specific version then?

lukasaric replied:
@MiroDojkic I think this link points to the latest v2 API docs version...v3 is published and stable and it's prefixed with `v3/latest/...`

lukasaric replied:
@MiroDojkic also we should consider bumping the version to v3 in the future since it's stable and recommended for general usage (it's written in v2 docs 😄 ).

ikovac commented on server/shared/storage/amazon.ts:
```diff
@@ -0,0 +1,123 @@
+  }
+
+  // API docs: http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#putObject-property
+  async saveFile(key: string, data: Buffer): Promise<void> {
+    const params = { Bucket: this.#bucket, Key: key, Body: data };
+    await this.#client.putObject(params).promise();
+  }
```
```suggestion
  saveFile(key: string, data: Buffer): Promise<void> {
    const params = { Bucket: this.#bucket, Key: key, Body: data };
    return this.#client.putObject(params).promise();
  }
```
Can we get rid of async await here?

lukasaric replied:
No, we can't, since the function return type is `Promise<void>` and `putObject` return type differs from this one.

ikovac commented on server/shared/storage/amazon.ts:
```diff
@@ -0,0 +1,123 @@
+  async deleteFile(key: string): Promise<DeleteResponse> {
+    const params = { Bucket: this.#bucket, Key: key };
+    await this.#client.deleteObject(params).promise();
+    return { isDeleted: true };
```
Do we also need to cast provider's errors to ours so that errors are consistent between providers? It is a little bit of work so I am not sure about this one, @MiroDojkic what do you think?

lukasaric replied:
I agree. I think we can leave this as it is, but definitely, we should implement some kind of united Error handler for all providers in the future.
Fly drive handles that by having unified [exceptions](https://github.com/Slynova-Org/flydrive/tree/develop/packages/flydrive/src/exceptions). I think we should do something like this...

MiroDojkic replied:
@ikovac makes sense to me, but we can leave that for improvement PR since we didn't have that interface previously.

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/storage-abstraction-enhancement

MiroDojkic commented on server/shared/storage/amazon.ts:
```diff
@@ -0,0 +1,123 @@
+
+    this.#bucket = amazonConfig.bucket;
+    this.#client = new S3(s3Config);
+  }
```
Consider whether this is simpler to read/less indirection:
```suggestion
  constructor(config: Config) {
    this.#bucket = config.storage.amazon.bucket;
    this.#client = new S3({
      accessKeyId: config.storage.amazon.key,
      secretAccessKey: config.storage.amazon.secret,
      region: config.storage.amazon.region,
      signatureVersion: 'v4',
      apiVersion: '2006-03-01',
      maxRetries: 3
    });
  }
```
or some similar variation.
Basically `config.storage.amazon` is S3 config so having `amazonConfig` and `s3Config` seems misleading.

MiroDojkic commented on server/shared/storage/amazon.ts:
```diff
@@ -0,0 +1,123 @@
+  getFile(key: string): Promise<ContentResponse<string>> {
+    const params = { Bucket: this.#bucket, Key: key };
+    return this.#client.getObject(params).promise()
+      .then(({ Body: data }) => ({ content: data.toString(), raw: data as Buffer }))
```
Do we need `data as Buffer` cast here?

lukasaric replied:
Data response here is typed as `S3.Body`, so buffer casting is needed here.

MiroDojkic commented on server/shared/storage/amazon.ts:
```diff
@@ -0,0 +1,123 @@
+  createWriteStream(key: string): NodeJS.WritableStream {
+    const throughStream = miss.through();
+    const params = { Bucket: this.#bucket, Key: key, Body: throughStream };
+    this.#client.upload(params, () => null);
```
Consider aliasing `noop` to reduce noise.

lukasaric replied:
As Erik Meijer said: "There to much noise" 😄

MiroDojkic commented on server/shared/storage/amazon.ts:
```diff
@@ -0,0 +1,123 @@
+      Delete: { Objects: objects }
+    };
+    await this.#client.deleteObjects(params).promise();
+    return { isDeleted: true };
```
Should we keep the interface consistent and return `Promise<void>` here?

lukasaric replied:
👍

MiroDojkic commented on server/shared/storage/amazon.ts:
```diff
@@ -0,0 +1,123 @@
+    const params = { Bucket: this.#bucket, Key: key };
+    return this.#client.headObject(params).promise()
+      .then(Boolean)
+      .catch(() => false);
```
Should we match against `NotFound` and rest to enable specific error handling (i.e. if file exists but the access is forbidden)?

MiroDojkic commented on server/shared/storage/interface.ts:
```diff
@@ -0,0 +1,21 @@
+interface IStorage {
+  getFile(key: string): Promise<ContentResponse<string>>;
+  createReadStream(key: string): NodeJS.ReadableStream;
+  createWriteStream(key: string): Promise<NodeJS.WritableStream> | NodeJS.WritableStream;
```
Can we ensure this method has consistent return - `Promise<NodeJS.WritableStream>`?

lukasaric replied:
👍

lukasaric committed:
Addressed issues :lipstick:

- Resolved provider path within the cr8 storage fn
- Set return type for file deletion to be Promise<void>
- Removed DeleteResponse type
- Amazon config setup in ctr adjustments
- Extracted noop fn
- Aligned writeStream creation to be resolved as promise

lukasaric committed:
isNotFound error improvement

lukasaric committed:
Export providers within the index

lukasaric committed:
Remove empty line

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/storage-abstraction-enhancement

lukasaric committed:
Put API docs above the s3 init

lukasaric committed:
Merge branch 'chore/migrate-to-typescript-with-bottlejs-and-mikro-orm' into feature/storage-abstraction-enhancement
